src/ir/drop.cpp

```diff
@@ -64,7 +64,12 @@ Expression* getDroppedChildrenAndAppend(Expression* curr,
 
   std::vector<Expression*> contents;
   for (auto* child : ChildIterator(curr)) {
-    if (!EffectAnalyzer(options, wasm, child).hasUnremovableSideEffects()) {
+    EffectAnalyzer effects(options, wasm, child);
+    // Ignore a trap, as the unreachable replacement would trap too.
+    if (last->type == Type::unreachable) {
```

Comment:
Interesting idea, but I think there are some corner cases here:

* We should check `last->is<Unreachable>()` as e.g. `(br $foo)` would have type unreachable but not trap.
* We can reorder traps but not with other effects. For example if child0 has a possible trap, child1 does a call, and then we append a trap, we can't remove child0 - we need to run that trap before doing the call, which might have noticeable effects.

I think this can still be done, but I'm not sure how much more complex handling that last situation is?

Reply:
Hmm I see, it's more complicated that I expected..

> * We should check `last->is<Unreachable>()` as e.g. `(br $foo)` would have type unreachable but not trap.

The is what we are already doing, which checks only if the type is unreachable, fine? https://github.com/WebAssembly/binaryen/blob/c40b402aaffa24c641cbb0e8efd0019c5c4a68b6/src/ir/drop.cpp#L46-L49

> * We can reorder traps but not with other effects. For example if child0 has a possible trap, child1 does a call, and then we append a trap, we can't remove child0 - we need to run that trap before doing the call, which might have noticeable effects.
> I think this can still be done, but I'm not sure how much more complex handling that last situation is?

I see, this is tricky... This probably is not worth it then. Thanks.

Reply:
> The is what we are already doing, which checks only if the type is unreachable, fine? 

Good point, yes, that code doesn't look correct. It should check `is<Unreachable>()`. Do you want to open a PR?

Reply:
Opened #4834.